storage/series/buffer.go

```diff
@@ -660,20 +688,75 @@ func (b *dbBufferBucket) discardMerged() discardMergedResult {
 		}
 	}
 
+	var lastWriteAt time.Time
 	iter := b.opts.MultiReaderIteratorPool().Get()
 	iter.Reset(readers)
+
+	defer func() {
+		iter.Close()
+		for _, stream := range readers {
+			if resource, ok := stream.(context.Finalizer); ok {
+				resource.Finalize()
+			}
+		}
+	}()
+
 	for iter.Next() {
 		dp, unit, annotation := iter.Current()
 		if err := encoder.Encode(dp, unit, annotation); err != nil {
-			log := b.opts.InstrumentOptions().Logger()
-			log.Errorf("buffer merge encode error: %v", err)
-			break
+			return mergeResult{}, err
 		}
+		lastWriteAt = dp.Timestamp
+	}
```

Comment:
do you need to check `iter.Err()` after this?

Reply:
Good catch, cheers.